### PR TITLE
Filter and search

### PR DESCRIPTION
**Pull request by ddjnw1yu**
This pull request contains:
1. New filter-related support api, reduces the loading time.
2. Irods search-related api, provide keywords search feature. The result will be ordered by relevance.
3. Separated graphql apis based on different use.
4. Updated test cases with latest apis.

### CODE REVIEW

File: app/filter.py [hunk 1]
```diff
@@ -1,93 +1,117 @@
...
+    "MAPPED_FUNDING": {
+        "title": "FUNDINGS",
+        "node": "dataset_description",
+        "field": "funding",
+        "element": {
+            "Funding A": ["OT3OD025349"],
+            "Funding B": ["OT2OD023848"],
+            "Funding C": ["OT2OD023847"]}
+    },
+    "MAPPED_MIME_TYPES2": {
```
Comment (alan-wu) [inline app/filter.py]: Do we need this one?

File: app/filter.py [hunk 2]
```diff
@@ -1,93 +1,117 @@
...
+            # "SCAFFOLD_DIR": ["inode/vnd.abi.scaffold+directory"],
+            "Plot2": ["text/vnd.abi.plot+tab-separated-values", "text/vnd.abi.plot+csv"],
+            # "COMMON_IMAGES": ["image/png", "image/jpeg"],
+            # "tiff-image": ["image/tiff", "image/tif"],
+            # "BIOLUCIDA_3D": ["image/jpx", "image/vnd.ome.xml+jpx"],
+            # "BIOLUCIDA_2D": ["image/jp2", "image/vnd.ome.xml+jp2"],
+            # "VIDEO": ["video/mp4"]
+        }
+    },
+    "MAPPED_FUNDING2": {
```
Comment (alan-wu) [inline app/filter.py]: Can we replace this one with species?

File: app/main.py [hunk 3]
```diff
@@ -296,102 +324,79 @@ async def get_gen3_record(uuid: str, item: Gen3Item):
...
+    """
+    Return the support data for frontend filters component.
+    """
+    return f.generate_filter_information()
+
+
+@ app.post("/filter/argument")
+async def get_filter_argument(item: GraphQLQueryItem):
+    query_result = graphql(item)
+    return f.generate_dataset_list(query_result[item.node])
```
Comment (alan-wu) [inline app/main.py]: Do these two routes do completely different things?

Probably can find a better name for the second route.

File: app/main.py [hunk 4]
```diff
@@ -291,107 +321,84 @@ async def get_gen3_record(uuid: str, item: Gen3Item):
...
+
+@ app.get("/filter")
+async def generate_filter():
+    """
+    Return the support data for frontend filters component.
+    """
+    return f.generate_filter_information()
+
+
+@ app.post("/filter/argument")
```
Comment (alan-wu) [inline app/main.py]: Would filter/dataset be a better route for this?
The function can be called get_filtered_datasets?

File: tests/test_api.py [hunk 5]
```diff
@@ -13,128 +13,201 @@ def test_get_gen3_program(client):
...
+        "program": "demo1",
+        "project": "12L",
+    }
+    response = client.post("/dictionary", json=pass_case)
     assert response.status_code == 200
 
+    missing_data = {}
+    response = client.post("/dictionary", json=missing_data)
+    assert response.status_code == 400
+    assert response.json()[
```
Comment (alan-wu) [inline tests/test_api.py]: Assign response.json() to a variable then access its element will be better for readability.

File: app/main.py [hunk 6]
```diff
@@ -450,18 +461,41 @@ class Config:
         }
 
 
+@ app.post("/search")
+async def search_content(item: SearchItem):
```
Comment (alan-wu) [inline app/main.py]: Please provide an example on the post request data. I would like to test it.

Comment (alan-wu) [reply app/main.py]: If it is an input string only, changing it to GET will be better,

**Review by alan-wu — CHANGES_REQUESTED** [review]
It looks good in general, I need to test the IRODS search but I am not able to do so at this moment. Thanks.

File: app/filter.py (hunk 2 above)
Comment (ddjnw1yu) [reply app/filter.py]: Sure, will do that now. It's just a test filter.

**Review by alan-wu — APPROVED** [review]

**Pull request merged**